src/utils/TypeGuards.ts

```diff
@@ -0,0 +1,184 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { IAttestation, ICredentialPresentation, IRequestCredentialContent } from '@kiltprotocol/types'
+
+import type {
+  IMessage,
+  IMessageBodyBase,
+  IRejectAttestation,
+  IRequestAttestation,
+  IRequestCredential,
+  ISubmitAttestation,
+  ISubmitCredential,
+  ISubmitTerms,
+} from '../types'
+
+export function isIMessage<Body extends IMessageBodyBase>(message: any): message is IMessage<Body> {
+  if (
+    typeof message !== 'object' ||
+    !('body' in message) ||
+    !('type' in message.body) ||
+    !('createdAt' in message) ||
+    !('sender' in message) ||
+    !('receiver' in message) ||
+    typeof message.createdAt !== 'number' ||
+    typeof message.sender !== 'string' ||
+    typeof message.receiver !== 'string'
+  ) {
+    return false
+  }
+
+  const { messageId, receivedAt, inReplyTo, references } = message
+
+  return (
+    (typeof messageId === 'undefined' || typeof messageId === 'string') &&
+    (typeof receivedAt === 'undefined' || typeof receivedAt === 'number') &&
+    (typeof inReplyTo === 'undefined' || typeof inReplyTo === 'string') &&
+    (Array.isArray(references) || typeof references === 'undefined')
+  )
+}
+
+export function isSubmitTerms(
+  message: IMessage<{ type: string; content: unknown }>
```

Comment:
tip: because you provide a default in the message typing, you don't need to repeat this here
```suggestion
  message: IMessage
```

Reply:
Ah yeah, in the first iteration I did not have the defaults. I still have to give everything a look from top 2 bottom :smile: